doc/source/user/basics.interoperability.rst

```diff
@@ -0,0 +1,352 @@
+ >>> buf = arr.__array_interface__
+ >>> buf
+ {'data': (140497590272032, False), 'strides': None, 'descr': [('', '<i8')], 'typestr': '<i8', 'shape': (4,), 'version': 3}
+ >>> buf['shape'] = (2, 2)
+ >>> w = wrapper()
+ >>> w.__array_interface__ = buf
+ >>> new_arr = np.array(w, copy=False)
+ >>> new_arr
+ array([[1, 2],
+        [3, 4]])
+
+We can check that ``arr`` and ``new_arr`` share the same data buffer:
+
+ >>> new_arr[0, 0] = 1000
+ >>> new_arr
+ array([[1000,    2],
+        [   3,    4]])
+ >>> arr
+ array([1000, 2, 3, 4])
+
+
+The ``__array__`` protocol
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``__array__`` protocol acts as a dispatch mechanism and ensures that any
+NumPy-like object (an array, any object exposing the array interface, an object
+whose ``__array__`` method returns an array or any nested sequence) that
+implements it can be used as a NumPy array. If possible, this will mean using
+``__array__`` to create a NumPy ndarray view of the array-like object.
+Otherwise, this copies the data into a new ndarray object. This is not optimal,
+as coercing arrays into ndarrays may cause performance problems or create the
+need for copies and loss of metadata.
+
+To see an example of a custom array implementation including the use of the
+``__array__`` protocol, see `Writing custom array containers
+<https://numpy.org/devdocs/user/basics.dispatch.html>`__.
+
+Operating on foreign objects without converting
+-----------------------------------------------
+
+Consider the following function.
+
+ >>> import numpy as np
+ >>> def f(x):
+ ...     return np.mean(np.exp(x))
+
+We can apply it to a NumPy ndarray object directly:
+
+ >>> x = np.array([1, 2, 3, 4])
+ >>> f(x)
+ 21.1977562209304
+
+We would like this function to work equally well with any NumPy-like array
+object. Some of this is possible today with various protocol mechanisms within
+NumPy.
+
+NumPy allows a class to indicate that it would like to handle computations in a
+custom-defined way through the following interfaces:
+
+-  ``__array_ufunc__``: allows third-party objects to support and override
+   :ref:`ufuncs <ufuncs-basics>`.
+-  ``__array_function__``: a catch-all for NumPy functionality that is not
+   covered by the ``__array_ufunc__`` protocol for universal functions.
+
+As long as foreign objects implement the ``__array_ufunc__`` or
+``__array_function__`` protocols, it is possible to operate on them without the
+need for explicit conversion.
+
+The ``__array_ufunc__`` protocol
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A :ref:`universal function (or ufunc for short) <ufuncs-basics>` is a
+“vectorized” wrapper for a function that takes a fixed number of specific inputs
+and produces a fixed number of specific outputs. The output of the ufunc (and
+its methods) is not necessarily an ndarray, if all input arguments are not
+ndarrays. Indeed, if any input defines an ``__array_ufunc__`` method, control
+will be passed completely to that function, i.e., the ufunc is overridden.
```

Comment:
I suggest adding a note here like: _the `__array_ufunc__` method on that (non-ndarray) object has access to the numpy ufunc. Because ufuncs have a well-defined structure, that `__array_ufunc__` method may rely on ufunc attributes like `.at()`, `.reduce()`, etc._